libp2p/transports/wstransport.nim

```diff
@@ -0,0 +1,199 @@
+
+type
+  WsStream = ref object of Connection
+    session: WSSession
+
+proc init*(T: type WsStream,
+           session: WSSession,
+           dir: Direction,
+           timeout = 10.minutes,
+           observedAddr: MultiAddress = MultiAddress()): T =
+
+  let stream = T(session: session,
+             timeout: timeout,
+             dir: dir,
+             observedAddr: observedAddr)
+  stream.initStream()
+  return stream
+
+method readOnce*(
+  s: WsStream,
+  pbytes: pointer,
+  nbytes: int):
+  Future[int] =
+  return s.session.recv(pbytes, nbytes)
+
+method write*(s: WsStream, msg: seq[byte]):
+  Future[void] {.async.} =
+  #TODO the write procs can raise LPStreamEOF
+  # BUT this is not indicated in .raises. bc of async
+  # So this function can't just return the Future of ws.send
+  # We should add {.raises.} to every write, but this will probably
+  # break many things
+  try:
+    await s.session.send(msg, Opcode.Binary)
+  except WSClosedError:
+    raise newLPStreamEOFError()
+
+method closeImpl*(s: WsStream): Future[void] {.async.} =
+  await s.session.close()
+  await procCall Connection(s).closeImpl()
+
+type
+  WsTransport* = ref object of Transport
+    httpserver: HttpServer
+    wsserver: WSServer
+    outconnections: seq[WsStream]
+
+    tlsPrivateKey: TLSPrivateKey
+    tlsCertificate: TLSCertificate
+
+
+method start*(
+  self: WsTransport,
+  ma: MultiAddress) {.async.} =
+  ## listen on the transport
+  ##
+
+  if self.running:
+    trace "WS transport already running"
+    return
+
+  await procCall Transport(self).start(ma)
+  trace "Starting WS transport"
+
+  self.httpserver =
+    if isNil(self.tlsPrivateKey):
+      HttpServer.create(self.ma.initTAddress().tryGet())
+    else:
+      TlsHttpServer.create(self.ma.initTAddress().tryGet(),
+        self.tlsPrivateKey,
+        self.tlsCertificate)
+
+  self.wsserver = WSServer.new()
+
+  # always get the resolved address in case we're bound to 0.0.0.0:0
+  self.ma = MultiAddress.init(self.httpserver.sock.getLocalAddress()).tryGet()
+  self.running = true
+
+  trace "Listening on", address = self.ma
+
+method stop*(self: WsTransport) {.async, gcsafe.} =
+  ## stop the transport
+  ##
+
+  self.running = false # mark stopped as soon as possible
+
+  try:
+    trace "Stopping WS transport"
+    await procCall Transport(self).stop() # call base
+
+    # server can be nil
+    if not isNil(self.httpserver):
+      self.httpserver.stop()
+      await self.httpserver.closeWait()
+
+    for conn in self.outconnections:
+      await conn.close()
+
+    self.httpserver = nil
+    trace "Transport stopped"
+  except CatchableError as exc:
+    trace "Error shutting down ws transport", exc = exc.msg
+
+method accept*(self: WsTransport): Future[Connection] {.async, gcsafe.} =
+  ## accept a new WS connection
+  ##
+
+  if not self.running:
+    raise newTransportClosedError()
+
+  try:
+    let
+      transp = await self.httpserver.accept()
```

Comment:
Lots of potential to unify with tcp transport `accept`, specially after we've enabled multiple address listening.

Reply:
Yep, when #598 will be more stable I'll rebase this PR on it and add the required changes :)